go/parquet/internal/encoding/memo_table_test.go

```diff
@@ -0,0 +1,284 @@
+)
+
+type MemoTableTestSuite struct {
+	suite.Suite
+}
+
+func TestMemoTable(t *testing.T) {
+	suite.Run(t, new(MemoTableTestSuite))
+}
+
+func (m *MemoTableTestSuite) assertGetNotFound(table encoding.MemoTable, v interface{}) {
+	_, ok := table.Get(v)
+	m.False(ok)
+}
+
+func (m *MemoTableTestSuite) assertGet(table encoding.MemoTable, v interface{}, expected int) {
+	idx, ok := table.Get(v)
+	m.Equal(expected, idx)
+	m.True(ok)
+}
+
+func (m *MemoTableTestSuite) assertGetOrInsert(table encoding.MemoTable, v interface{}, expected int) {
+	idx, _, err := table.GetOrInsert(v)
+	m.NoError(err)
+	m.Equal(expected, idx)
+}
+
+func (m *MemoTableTestSuite) assertGetNullNotFound(table encoding.MemoTable) {
+	_, ok := table.GetNull()
+	m.False(ok)
+}
+
+func (m *MemoTableTestSuite) assertGetNull(table encoding.MemoTable, expected int) {
+	idx, ok := table.GetNull()
+	m.Equal(expected, idx)
+	m.True(ok)
+}
+
+func (m *MemoTableTestSuite) assertGetOrInsertNull(table encoding.MemoTable, expected int) {
+	idx, _ := table.GetOrInsertNull()
+	m.Equal(expected, idx)
+}
+
+func (m *MemoTableTestSuite) TestInt64() {
+	const (
+		A int64 = 1234
+		B int64 = 0
+		C int64 = -98765321
+		D int64 = 12345678901234
+		E int64 = -1
+		F int64 = 1
+		G int64 = 9223372036854775807
+		H int64 = -9223372036854775807 - 1
+	)
+
+	// table := encoding.NewInt64MemoTable(nil)
+	table := hashing.NewInt64MemoTable(0)
+	m.Zero(table.Size())
+	m.assertGetNotFound(table, A)
+	m.assertGetNullNotFound(table)
+	m.assertGetOrInsert(table, A, 0)
+	m.assertGetNotFound(table, B)
+	m.assertGetOrInsert(table, B, 1)
+	m.assertGetOrInsert(table, C, 2)
+	m.assertGetOrInsert(table, D, 3)
+	m.assertGetOrInsert(table, E, 4)
+	m.assertGetOrInsertNull(table, 5)
+
+	m.assertGet(table, A, 0)
+	m.assertGetOrInsert(table, A, 0)
+	m.assertGet(table, E, 4)
+	m.assertGetOrInsert(table, E, 4)
+
+	m.assertGetOrInsert(table, F, 6)
+	m.assertGetOrInsert(table, G, 7)
+	m.assertGetOrInsert(table, H, 8)
+
+	m.assertGetOrInsert(table, G, 7)
+	m.assertGetOrInsert(table, F, 6)
+	m.assertGetOrInsertNull(table, 5)
+	m.assertGetOrInsert(table, E, 4)
+	m.assertGetOrInsert(table, D, 3)
+	m.assertGetOrInsert(table, C, 2)
+	m.assertGetOrInsert(table, B, 1)
+	m.assertGetOrInsert(table, A, 0)
+
+	const sz int = 9
+	m.Equal(sz, table.Size())
+	m.Panics(func() {
+		values := make([]int32, sz)
+		table.CopyValues(values)
+	}, "should panic because wrong type")
+	m.Panics(func() {
+		values := make([]int64, sz-3)
+		table.CopyValues(values)
+	}, "should panic because out of bounds")
+
+	{
+		values := make([]int64, sz)
+		table.CopyValues(values)
+		m.Equal([]int64{A, B, C, D, E, 0, F, G, H}, values)
+	}
+	{
+		const offset = 3
+		values := make([]int64, sz-offset)
+		table.CopyValuesSubset(offset, values)
+		m.Equal([]int64{D, E, 0, F, G, H}, values)
+	}
+}
+
+func (m *MemoTableTestSuite) TestFloat64() {
+	const (
+		A float64 = 0.0
+		B float64 = 1.5
+		C float64 = -0.1
+	)
+	var (
+		D = math.Inf(1)
+		E = -D
+		F = math.NaN()
```

Comment:
ideally, you would test with to different bit representations of Nan

Reply:
That's a good point, i'll look up a couple different bit representations of Nan to compare with and add them in.

Reply:
added tests with 2 other bit representations of NaN for comparisons